hazelcast/src/main/java/com/hazelcast/sql/impl/exec/io/ReceiveSortMergeExec.java

```diff
@@ -0,0 +1,225 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.exec.io;
+
+import com.hazelcast.sql.impl.exec.AbstractExec;
+import com.hazelcast.sql.impl.exec.IterationResult;
+import com.hazelcast.sql.impl.exec.sort.MergeSort;
+import com.hazelcast.sql.impl.exec.sort.MergeSortSource;
+import com.hazelcast.sql.impl.exec.sort.SortKey;
+import com.hazelcast.sql.impl.exec.sort.SortKeyComparator;
+import com.hazelcast.sql.impl.expression.Expression;
+import com.hazelcast.sql.impl.row.EmptyRowBatch;
+import com.hazelcast.sql.impl.row.ListRowBatch;
+import com.hazelcast.sql.impl.row.Row;
+import com.hazelcast.sql.impl.row.RowBatch;
+import com.hazelcast.sql.impl.worker.QueryFragmentContext;
+
+import java.util.List;
+
+/**
+ * Executor which receives entries from multiple sources and merges them into a single sorted stream.
+ */
+@SuppressWarnings("rawtypes")
+public class ReceiveSortMergeExec extends AbstractExec {
+    /**
+     * AbstractInbox to consume results from.
+     */
+    private final StripedInbox inbox;
+
+    /**
+     * Expressions.
+     */
+    private final List<Expression> expressions;
+
+    /**
+     * Sorter.
+     */
+    private final MergeSort sorter;
+
+    /**
+     * Current batch.
+     */
+    private RowBatch curBatch;
+
+    public ReceiveSortMergeExec(
+        int id,
+        StripedInbox inbox,
+        List<Expression> expressions,
+        List<Boolean> ascs
+    ) {
+        super(id);
+
+        this.inbox = inbox;
+        this.expressions = expressions;
+
+        MergeSortSource[] sources = new MergeSortSource[inbox.getStripeCount()];
+
+        for (int i = 0; i < inbox.getStripeCount(); i++) {
+            sources[i] = new Source(i);
+        }
+
+        sorter = new MergeSort(sources, new SortKeyComparator(ascs));
+    }
+
+    @Override
+    protected void setup0(QueryFragmentContext ctx) {
+        inbox.setup();
+    }
+
+    @Override
+    public IterationResult advance0() {
+        while (true) {
```

Comment:
Static code analysis indicates that `while(true)` actually never loops. We need isolated unit tests for all involved components (`ReceiveSortMergeExec`, `MergeSort`, `MergeSortSource`, `SortKey`, `SortKeyComparator`). Otherwise, there is a high risk of bugs, since these classes are complex.

Reply:
Removed unnecessary loop in https://github.com/hazelcast/hazelcast/pull/18013/commits/dc96300140c57bcb405172ff5f4aa3e48724d499
The tests will be in the follow-up commit.

Reply:
Added tests in https://github.com/hazelcast/hazelcast/pull/18013/commits/132d767c2a5c783fe2fc0d693609a5d9ce7ac499